experimental/metal/metal_driver.m

```diff
@@ -0,0 +1,450 @@
+  iree_hal_device_info_t* device_infos = NULL;
+  iree_host_size_t single_info_size =
+      sizeof(iree_hal_device_info_t) +
+      (IREE_HAL_METAL_MAX_DEVICE_PATH_LENGTH + IREE_HAL_METAL_MAX_DEVICE_NAME_LENGTH) *
+          sizeof(char);
+  iree_host_size_t total_size = device_count * single_info_size;
+  iree_status_t status = iree_allocator_malloc(host_allocator, total_size, (void**)&device_infos);
+
+  if (iree_status_is_ok(status)) {
+    // Append all path and name strings at the end of the struct.
+    uint8_t* buffer_ptr = (uint8_t*)device_infos + device_count * sizeof(iree_hal_device_info_t);
+    for (iree_host_size_t i = 0; i < device_count; ++i) {
+      status = iree_hal_metal_populate_device_info(devices[i], buffer_ptr, &buffer_ptr,
+                                                   &device_infos[i]);
+      if (!iree_status_is_ok(status)) break;
+    }
+  }
+  if (iree_status_is_ok(status)) {
+    *out_device_info_count = device_count;
+    *out_device_infos = device_infos;
+  } else {
+    iree_allocator_free(host_allocator, device_infos);
+  }
+  return status;
+}
+
+// Returns the GPU family the given |device| supports. Returns 0 if the given |device| does not
+// belong to a GPU family considered by IREE right now.
+static MTLGPUFamily iree_hal_metal_apple_gpu_family_query(id<MTLDevice> device) {
+  // Inspect whether the given device is a specific Apple GPU.
+  if ([device supportsFamily:MTLGPUFamilyApple8]) return MTLGPUFamilyApple8;
+  if ([device supportsFamily:MTLGPUFamilyApple7]) return MTLGPUFamilyApple7;
+  if ([device supportsFamily:MTLGPUFamilyApple6]) return MTLGPUFamilyApple6;
+  if ([device supportsFamily:MTLGPUFamilyApple5]) return MTLGPUFamilyApple5;
+  if ([device supportsFamily:MTLGPUFamilyApple4]) return MTLGPUFamilyApple4;
+  if ([device supportsFamily:MTLGPUFamilyApple3]) return MTLGPUFamilyApple3;
+  if ([device supportsFamily:MTLGPUFamilyApple2]) return MTLGPUFamilyApple2;
+  if ([device supportsFamily:MTLGPUFamilyApple1]) return MTLGPUFamilyApple1;
+
+  // Inspect whether whether the given GPU falls into some common family.
+  if ([device supportsFamily:MTLGPUFamilyCommon3]) return MTLGPUFamilyCommon3;
+  if ([device supportsFamily:MTLGPUFamilyCommon2]) return MTLGPUFamilyCommon2;
+  if ([device supportsFamily:MTLGPUFamilyCommon1]) return MTLGPUFamilyCommon1;
+
+  return 0;
+}
+
+static const char* iree_hal_metal_get_gpu_family_name(MTLGPUFamily family) {
+  switch (family) {
+    case MTLGPUFamilyApple8:
+      return "apple8(a15/m2)";
+    case MTLGPUFamilyApple7:
+      return "apple7(a14/m1)";
+    case MTLGPUFamilyApple6:
+      return "apple6(a13)";
+    case MTLGPUFamilyApple5:
+      return "apple5(a12)";
+    case MTLGPUFamilyApple4:
+      return "apple4(a11)";
+    case MTLGPUFamilyApple3:
+      return "apple3(a9/a10)";
+    case MTLGPUFamilyApple2:
+      return "apple2(a8)";
+    case MTLGPUFamilyApple1:
+      return "apple1(a7)";
```

Comment:
Do we need all of these? Seems like some binary size bloat.

Reply:
I'd prefer to keep for the usefulness--just a few bytes more anyway. :)